google/resource_compute_instance_test.go

```diff
@@ -3881,3 +4104,178 @@ resource "google_compute_instance" "foobar" {
+
+	boot_disk {
+		initialize_params{
+			image = "${data.google_compute_image.my_image.self_link}"
+		}
+	}
+
+	network_interface {
+		network = "default"
+	}
+
+	metadata = {
+		foo = "bar"
+	}
+}
+`, instance)
+}
+
+func testAccComputeInstance_desiredStatusRunning(instance string) string {
+	return fmt.Sprintf(`
+data "google_compute_image" "my_image" {
+	family  = "debian-9"
+	project = "debian-cloud"
+}
+
+resource "google_compute_instance" "foobar" {
+	name           = "%s"
+	machine_type   = "n1-standard-1"
+	zone           = "us-central1-a"
+	can_ip_forward = false
+	tags           = ["foo", "bar"]
+
+	boot_disk {
+		initialize_params{
+			image = "${data.google_compute_image.my_image.self_link}"
+		}
+	}
+
+	network_interface {
+		network = "default"
+	}
+
+	desired_status = "RUNNING"
+
+	metadata = {
+		foo = "bar"
+	}
+}
+`, instance)
+}
+
+func testAccComputeInstance_desiredStatusTerminated(instance string) string {
+	return fmt.Sprintf(`
+data "google_compute_image" "my_image" {
+	family  = "debian-9"
+	project = "debian-cloud"
+}
+
+resource "google_compute_instance" "foobar" {
+	name           = "%s"
+	machine_type   = "n1-standard-1"
+	zone           = "us-central1-a"
+	can_ip_forward = false
+	tags           = ["foo", "bar"]
+
+	boot_disk {
+		initialize_params{
+			image = "${data.google_compute_image.my_image.self_link}"
+		}
+	}
+
+	network_interface {
+		network = "default"
+	}
+
+	desired_status = "TERMINATED"
+
+	metadata = {
+		foo = "bar"
+	}
+}
+`, instance)
+}
+
+func testAccComputeInstance_desiredStatusTerminatedUpdate(instance string) string {
+	return fmt.Sprintf(`
+data "google_compute_image" "my_image" {
+	family  = "debian-9"
+	project = "debian-cloud"
+}
+
+resource "google_compute_instance" "foobar" {
+	name           = "%s"
+	machine_type   = "n1-standard-1"
+	zone           = "us-central1-a"
+	can_ip_forward = false
+	tags           = ["baz"]
+
+	boot_disk {
+		initialize_params{
+			image = "${data.google_compute_image.my_image.self_link}"
+		}
+	}
+
+	network_interface {
+		network = "default"
+		access_config { }
+	}
+
+	desired_status = "TERMINATED"
+
+	metadata = {
+		bar = "baz"
+	}
+
+	labels = {
+		only_me = "nothing_else"
+	}
+}
+`, instance)
+}
+
+func testAccComputeInstance_desiredStatusTerminatedUpdateRequiringStopping(instance string) string {
+	return fmt.Sprintf(`
+data "google_compute_image" "my_image" {
+	family  = "debian-9"
+	project = "debian-cloud"
+}
+
+resource "google_compute_instance" "foobar" {
+	name           = "%s"
+	machine_type   = "n1-standard-2"
+	zone           = "us-central1-a"
+	can_ip_forward = false
+	tags           = ["baz"]
+
+	boot_disk {
+		initialize_params{
+			image = "${data.google_compute_image.my_image.self_link}"
+		}
+	}
+
+	network_interface {
+		network = "default"
+		access_config { }
+	}
+
+	desired_status = "TERMINATED"
+
+	metadata = {
+		bar = "baz"
+	}
+
+	labels = {
+		only_me = "nothing_else"
+	}
+
+	allow_stopping_for_update = true
```

Comment:
I'm not sure about this. If the instance has status `TERMINATED`, there should be no issue updating some fields requiring stopping the instance (here, the machine_type). I could add the following check : if `instance.Status = "TERMINATED"`, then the `allow_stopping_for_update` is not required. What do you think?

Reply:
I think we can skip checking `allow_stopping_for_update` if the instance is already terminated, or if the `desired_status` is `TERMINATED`. We'll treat it as "allow stopping+starting a RUNNING instance to perform an update".

Additionally, if the instance was RUNNING and we stopped it for an update, we should restore it to RUNNING if `desired_status` is `""` or `"RUNNING"`.

My expectations, assuming we run `terraform apply` changing the `machine_type`:

- a RUNNING instance with desired_status of `""` or `"RUNNING"` and `allow_stopping_for_update = true` will be updated and restored to RUNNING

- a RUNNING instance with desired_status of `""` or `"RUNNING"` and `allow_stopping_for_update = false` will fail to update

- a RUNNING instance with desired_status of `"TERMINATED"` with `allow_stopping_for_update = false` can be stopped early to allow updates

- a TERMINATED instance with desired_status of `"TERMINATED"` with `allow_stopping_for_update = false` is already stopped, and will remain stopped

- a TERMINATED instance with desired_status of `""` with `allow_stopping_for_update = false` is already stopped, and will remain stopped

- a TERMINATED instance with desired_status of `"RUNNING"` with `allow_stopping_for_update = false` is updated while stopped, and then started

Let me know if you'd expect different behaviour or if you feel there's an interesting case that isn't covered.

Reply:
Thanks for the details you provided, I made a table to summarize all possible cases. Each case is covered by an acceptance test :

```                                                                                                 
+----------------------------------------------------------------------+---------+-----------------------+-------------------------------------------------------------------------------------------+
| Input                                                                |         | Actions (ordered)     |                                                                                           |
+----------------------------------------------------------------------+         +-----------------------+                                                                                           |
| Current instance status | desired_status | allow_stopping_for_update | Success | Stop | Update | Start | Acceptance test                                                                           |
+-------------------------+----------------+---------------------------+---------+------+--------+-------+-------------------------------------------------------------------------------------------+
| RUNNING                 | ""             | TRUE                      | X       | X    | X      | X     | TestAccComputeInstance_updateRunning_desiredStatusNotSet_allowStoppingForUpdate           |
| RUNNING                 | RUNNING        | TRUE                      | X       | X    | X      | X     | TestAccComputeInstance_updateRunning_desiredStatusRunning_allowStoppingForUpdate          |
| RUNNING                 | ""             | FALSE                     |         |      |        |       | TestAccComputeInstance_updateRunning_desiredStatusNotSet_notAllowStoppingForUpdate        |
| RUNNING                 | RUNNING        | FALSE                     |         |      |        |       | TestAccComputeInstance_updateRunning_desiredStatusRunning_notAllowStoppingForUpdate       |
| RUNNING                 | TERMINATED     | TRUE                      | X       | X    | X      |       | TestAccComputeInstance_updateRunning_desiredStatusTerminated_allowStoppingForUpdate       |
| RUNNING                 | TERMINATED     | FALSE                     | X       | X    | X      |       | TestAccComputeInstance_updateRunning_desiredStatusTerminated_notAllowStoppingForUpdate    |
+-------------------------+----------------+---------------------------+---------+------+--------+-------+-------------------------------------------------------------------------------------------+
| TERMINATED              | ""             | TRUE                      | X       |      | X      |       | TestAccComputeInstance_updateTerminated_desiredStatusNotSet_allowStoppingForUpdate        |
| TERMINATED              | TERMINATED     | TRUE                      | X       |      | X      |       | TestAccComputeInstance_updateTerminated_desiredStatusTerminated_allowStoppingForUpdate    |
| TERMINATED              | ""             | FALSE                     | X       |      | X      |       | TestAccComputeInstance_updateTerminated_desiredStatusNotSet_notAllowStoppingForUpdate     |
| TERMINATED              | TERMINATED     | FALSE                     | X       |      | X      |       | TestAccComputeInstance_updateTerminated_desiredStatusTerminated_notAllowStoppingForUpdate |
| TERMINATED              | RUNNING        | TRUE                      | X       |      | X      | X     | TestAccComputeInstance_updateTerminated_desiredStatusRunning_allowStoppingForUpdate       |
| TERMINATED              | RUNNING        | FALSE                     | X       |      | X      | X     | TestAccComputeInstance_updateTerminated_desiredStatusRunning_notAllowStoppingForUpdate    |
+-------------------------+----------------+---------------------------+---------+------+--------+-------+-------------------------------------------------------------------------------------------+
```

All acceptance tests are passing :smile: